### PR TITLE
Feature/beanstalk

### PR DESCRIPTION
**Pull request by sushant-115**
Fixes https://github.com/practo/k8s-worker-pod-autoscaler/issues/47

**Add beanstalkd support**
```
apiVersion: k8s.practo.dev/v1alpha1
kind: WorkerPodAutoScaler
metadata:
  name: example-wpa
spec:
  minReplicas: 0
  maxReplicas: 10
  targetMessagesPerWorker: 1
  deploymentName: example-deployment
  queueURI: bs|test-tube|127.0.0.1|11300
```
*Add CMD line argument for beanstalk poll interval*
```
command:
  - /workerpodautoscaler
  - run
  - --resync-period=20
  - --wpa-threads=10
  - --aws-regions=ap-south-1
  - --sqs-short-poll-interval=20
  - --sqs-long-poll-interval=20
  - --beanstalk-poll-interval=3
```

### CODE REVIEW

**Comment by alok87** [comment]
@sushant-115 Thanks for the beanstalk support.

Few things to do before we can go ahead with merging this.
- [ ] Please back merge with https://github.com/practo/k8s-worker-pod-autoscaler/pull/55, many dependencies including k8s 1.15+ and aws dependencies were updated.
- [ ] Please rebase.
- [ ] Code review by @justjkk 
- [ ] Test in practodev staging @alok87

**Comment by sushant-115** [comment]
@alok87 First two steps are done

File: go.mod [hunk 1]
```diff
@@ -3,16 +3,23 @@ module github.com/practo/k8s-worker-pod-autoscaler
...
-	k8s.io/apiextensions-apiserver v0.0.0
-	k8s.io/apimachinery v0.0.0
-	k8s.io/client-go v0.0.0
-	k8s.io/code-generator v0.0.0-20190612205613-18da4a14b22b
+	golang.org/x/tools v0.0.0-20200226224502-204d844ad48d // indirect
+	k8s.io/api v0.0.0-20190726022912-69e1bce1dad5
+	k8s.io/apiextensions-apiserver v0.0.0-20190726024412-102230e288fd
+	k8s.io/apimachinery v0.0.0-20190726022757-641a75999153
+	k8s.io/client-go v0.0.0-20190726023111-a9c895e7f2ac
+	k8s.io/code-generator v0.0.0-20190726022633-14ba7d03f06f
```
Comment (alok87) [inline go.mod]: @sushant-115 i don't think you back merged with #55

File: cmd/workerpodautoscaler/run.go [hunk 2]
```diff
@@ -59,7 +60,7 @@ func (v *runCmd) new() *cobra.Command {
 	flags.String("kube-config", "", "path of the kube config file, if not specified in cluster config is used")
 	flags.Int("sqs-short-poll-interval", 20, "the duration (in seconds) after which the next sqs api call is made to fetch the queue length")
 	flags.Int("sqs-long-poll-interval", 20, "the duration (in seconds) for which the sqs receive message call waits for a message to arrive")
-
+	flags.Int("beanstalk-poll-interval", 3, "the duration (in seconds) for which the beanstalk receive message call waits for a message to arrive")
```
Comment (alok87) [inline cmd/workerpodautoscaler/run.go]: need to have a flag to choose which all queue provider the WPA should have, a list like 
`queue-providers`.  And we should create and start the pollers for only those queue providers. Because some people would just want to run SQS and some just beanstalk and some might want both.

Comment (sushant-115) [reply cmd/workerpodautoscaler/run.go]: It sounds good. Sure will implement this.

File: go.mod (hunk 1 above)
Comment (sushant-115) [reply go.mod]: These were automatically updated by running `make`

Comment (alok87) [reply go.mod]: Its reverting the upgrades done on #55

Comment (sushant-115) [reply go.mod]: Done

File: cmd/workerpodautoscaler/run.go (hunk 2 above)
Comment (sushant-115) [reply cmd/workerpodautoscaler/run.go]: Implemented this

File: pkg/queue/queue.go [hunk 3]
```diff
@@ -179,6 +179,11 @@ func (q *Queues) GetQueueInfo(namespace string, name string) (string, int32, int
 }
 
 func parseQueueURI(uri string) (string, string, error) {
+
+	if strings.Contains(uri, "bs") {
```
Comment (justjkk) [inline pkg/queue/queue.go]: This is not required if queueURI is a valid URL with scheme `beanstalk`. Eg: `beanstalk://127.0.0.1:11300/test-tube`

This will match false-positives for cases like `https://sqs.ap-south-1.amazonaws.com/1234567890/test-bs-tube`

File: pkg/queue/queue.go [hunk 4]
```diff
@@ -207,8 +212,15 @@ func getProvider(host string, protocol string) (bool, string, error) {
...
 
 func getQueueName(name string) string {
-	splitted := strings.Split(name, "/")
-	return splitted[len(splitted)-1]
+	queueName := ""
+	if strings.Contains(name, "sqs") {
+		splitted := strings.Split(name, "/")
+		queueName = splitted[len(splitted)-1]
+	} else if strings.Contains(name, "bs") {
+		splitted := strings.Split(name, "|")
```
Comment (justjkk) [inline pkg/queue/queue.go]: Please refer to the other comment on using URL format.

File: pkg/queue/queue.go (hunk 3 above)
Comment (justjkk) [reply pkg/queue/queue.go]: I was testing the PR and hence I made the changes in a new branch and gave a PR [here](https://github.com/Exotel-Org/k8s-worker-pod-autoscaler/pull/1).

File: cmd/workerpodautoscaler/run.go [hunk 5]
```diff
@@ -33,6 +33,11 @@ var (
 	runExample = `  workerpodautoscaler run`
 )
 
+const (
+	sqsQueueService       = "sqs"
+	beanstalkQueueService = "beanstalkd"
+)
```
Comment (alok87) [inline cmd/workerpodautoscaler/run.go]: Please move these constants to https://github.com/practo/k8s-worker-pod-autoscaler/blob/master/pkg/queue/queueing_service.go

**Comment by alok87** [comment]
@sushant-115 Only 3 changes left and we are good to go live 🤘 
- [ ] Go mod tidy to fix go.sum
- [ ] Beanstalk URL fixes
- [ ] Queueing name and constants movement
- [ ] Documentation mentioning beanstalk. README has everything about SQS only.

File: pkg/queue/sqs.go [hunk 6]
```diff
@@ -81,7 +80,11 @@ func (s *SQS) getCWClient(queueURI string) *cloudwatch.CloudWatch {
 }
 
 func (s *SQS) longPollReceiveMessage(queueURI string) (int32, error) {
-	result, err := s.getSQSClient(queueURI).ReceiveMessage(&sqs.ReceiveMessageInput{
+	sqsClient := s.getSQSClient(queueURI)
+	if sqsClient == nil {
+		return 0, fmt.Errorf("Unable to fetch queue, check queue URL or permission")
```
Comment (alok87) [inline pkg/queue/sqs.go]: "Unable to make sqs client'

File: pkg/queue/queue.go (hunk 3 above)
Comment (sushant-115) [reply pkg/queue/queue.go]: Thanks. Merged it

File: cmd/workerpodautoscaler/run.go (hunk 5 above)
Comment (sushant-115) [reply cmd/workerpodautoscaler/run.go]: Moved this to queuing_service.go

File: pkg/queue/poller.go [hunk 7]
```diff
@@ -63,6 +63,7 @@ func (p *Poller) sync(stopCh <-chan struct{}) {
 	for {
 		select {
 		case listResultCh := <-p.listThreadCh:
+			time.Sleep(1 * time.Second)
```
Comment (alok87) [inline pkg/queue/poller.go]: Why was this added? Please suggest.

File: pkg/queue/sqs.go [hunk 8]
```diff
@@ -276,7 +295,7 @@ func (s *SQS) poll(key string, queueSpec QueueSpec) {
 					queueSpec.name, err)
 				return
 			} else {
-				klog.Fatalf("Unable to receive message from queue %q, %v.",
+				klog.Errorf("Unable to receive message from queue %q, %v.",
```
Comment (alok87) [inline pkg/queue/sqs.go]: I prefer it crash in these scenarios so that we get KubePodCrashLoop alert. These errors when fails is easier to fix.

File: pkg/queue/sqs.go [hunk 9]
```diff
@@ -296,19 +315,13 @@ func (s *SQS) poll(key string, queueSpec QueueSpec) {
 				queueSpec.name, err)
 			return
 		} else {
-			klog.Fatalf("Unable to get approximate messages in queue %q, %v.",
+			klog.Errorf("Unable to get approximate messages in queue %q, %v.",
```
Comment (alok87) [inline pkg/queue/sqs.go]: Same as https://github.com/practo/k8s-worker-pod-autoscaler/pull/57/files#r393239908

File: pkg/queue/sqs.go [hunk 10]
```diff
@@ -296,19 +315,13 @@ func (s *SQS) poll(key string, queueSpec QueueSpec) {
...
 		}
 	}
 	klog.Infof("%s: approxMessages=%d", queueSpec.name, approxMessages)
 	s.queues.updateMessage(key, approxMessages)
 
-	if approxMessages != 0 {
-		s.queues.updateIdleWorkers(key, -1)
-		s.waitForShortPollInterval()
-		return
-	}
```
Comment (alok87) [inline pkg/queue/sqs.go]: Why was this change made? Please revert them. The changes to SQS code should not be part of this PR (if a change is to be made for them please raise a separate PR)

File: pkg/queue/sqs.go [hunk 11]
```diff
@@ -341,12 +353,12 @@ func (s *SQS) poll(key string, queueSpec QueueSpec) {
 	// to a lower value even if it is possible
 	emptyReceives, err := s.cachedNumberOfEmptyReceives(queueSpec.uri)
 	if err != nil {
-		klog.Fatalf("Unable to fetch empty receive metric for queue %q, %v.",
+		klog.Errorf("Unable to fetch empty receive metric for queue %q, %v.",
```
Comment (alok87) [inline pkg/queue/sqs.go]: Same as https://github.com/practo/k8s-worker-pod-autoscaler/pull/57/files#r393240758

File: pkg/queue/sqs.go [hunk 12]
```diff
@@ -341,12 +353,12 @@ func (s *SQS) poll(key string, queueSpec QueueSpec) {
...
 			queueSpec.name, err)
 	}
 
 	var idleWorkers int32
-	if emptyReceives == 1.0 {
+	if emptyReceives == 1.0 && approxMessagesNotVisible <= 0 {
```
Comment (alok87) [inline pkg/queue/sqs.go]: Same as https://github.com/practo/k8s-worker-pod-autoscaler/pull/57/files#r393240758

File: pkg/queue/sqs.go (hunk 11 above)
Comment (alok87) [inline pkg/queue/sqs.go]: Same as https://github.com/practo/k8s-worker-pod-autoscaler/pull/57/files#r393240117

File: pkg/controller/controller.go [hunk 13]
```diff
@@ -370,29 +370,25 @@ func (c *Controller) getDesiredWorkers(
...
 			return currentWorkers
 		}
 
-		desiredWorkers := int32(math.Ceil(usageRatio * float64(currentWorkers)))
+		desiredWorkers := int32(math.Ceil(usageRatio))
 		// to prevent scaling down of workers which could be doing processing
-		if desiredWorkers < currentWorkers {
-			return currentWorkers
-		}
+
```
Comment (alok87) [inline pkg/controller/controller.go]: Why are we changing the formula here.. Please suggest. We should not be ideally changing these code in this PR. Please raise a separate PR with the reason mentioned.

File: pkg/controller/controller.go [hunk 14]
```diff
@@ -370,29 +370,25 @@ func (c *Controller) getDesiredWorkers(
 		return convertDesiredReplicasWithRules(desiredWorkers, minWorkers, maxWorkers)
 	}
 
-	if queueMessages > 0 {
+	if queueMessages > 0 && idleWorkers <= 0 {
```
Comment (alok87) [inline pkg/controller/controller.go]: Same as https://github.com/practo/k8s-worker-pod-autoscaler/pull/57/files#r393241930

**Comment by alok87** [comment]
Also please rebase(clean/squash commits) in the end after you have resolved all the review comments.

**Comment by alok87** [comment]
@sushant-115 can I also contribute to this? We are running this in our staging and facing some concurrent write issues also.

**Comment by alok87** [comment]
moved here #69
